File: src/test/java/com/libraryjava/service/UserServiceTest.java
package com.libraryjava.service;

import com.libraryjava.domain.user.User;
import com.libraryjava.domain.user.loanhistory.UserLoanHistory;
import com.libraryjava.domain.user.loanhistory.UserLoanStatus;
import com.libraryjava.dto.user.UserMakeDto;
import com.libraryjava.dto.user.UserResponseDto;
import com.libraryjava.dto.user.UserUpdateDto;
import com.libraryjava.dto.user.response.UserLoanHistoryResponse;
import com.libraryjava.repository.UserLoanHistoryRepository;
import com.libraryjava.repository.UserRepository;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

import static org.assertj.core.api.Assertions.*;
import static org.junit.jupiter.api.Assertions.*;

@SpringBootTest
class UserServiceTest {
    @Autowired UserService userService;
    @Autowired UserRepository userRepository;
    @Autowired
    private UserLoanHistoryRepository userLoanHistoryRepository;

    @AfterEach
    void clean() {
        System.out.println("userServiceTest CLEAN");
        userRepository.deleteAll();
    }

    @Test
    void makeUserTest() {
        // given
        UserMakeDto request = new UserMakeDto("김진규", 1000);

        // when
        userService.makeUser(request);

        // then
        List<User> users = userRepository.findAll();
        assertThat(users).hasSize(1);
        assertThat(users.get(0).getName()).isEqualTo("김진규");
        assertThat(users.get(0).getAge()).isEqualTo(1000);
    }

    @Test
    void getUsersTest() {
        // given
        userRepository.saveAll(List.of(User.fixture("A", 10), User.fixture("B", 20)));

        // when
        List<UserResponseDto> result = userService.getUsers();

        // then
        assertThat(result).hasSize(2);
        assertThat(result).extracting("name").containsExactlyInAnyOrder("A", "B");
        assertThat(result).extracting("age").containsExactlyInAnyOrder(10, 20);
    }

    @Test
    void updateUserTest() {
        // given
        User savedUser = userRepository.save(User.fixture("A", 10));
        UserUpdateDto request = new UserUpdateDto(savedUser.getId(), "B");

        // when
        userService.updateUser(request);

        // then
        User findUser = userRepository.findAll().get(0);
        assertThat(findUser.getName()).isEqualTo("B");
    }

    @Test
    void deleteUserTest() {
        // given
        User savedUser = userRepository.save(User.fixture("A", 10));

        // when
        userService.deleteUser("A");

        // then
        assertThat(userRepository.findAll()).isEmpty();
    }

    @Test
    @DisplayName("대출 기록이 없는 유저도 응답에 포함된다.")
    void getUserLoanHistoriesTest1() {
        // given
        userRepository.save(User.fixture("A", 10));

        // when
        List<UserLoanHistoryResponse> result = userService.getUserLoanHistories();

        // then
        assertThat(result).hasSize(1);
        assertThat(result.get(0).name()).isEqualTo("A");
        assertThat(result.get(0).books()).isEmpty();
    }

    @Test
    @DisplayName("대출 기록이 많은 유저의 응답이 정상 동작한다.")
    void getUserLoanHistoriesTest2() {
        // given
        User savedUser = userRepository.save(User.fixture("A", 10));
        userLoanHistoryRepository.saveAll(List.of(
                new UserLoanHistory("책1", UserLoanStatus.LOANED, savedUser),
                new UserLoanHistory("책2", UserLoanStatus.LOANED, savedUser),
                new UserLoanHistory("책3", UserLoanStatus.RETURNED, savedUser)
        ));

        // when
        List<UserLoanHistoryResponse> result = userService.getUserLoanHistories();

        // then
        assertThat(result).hasSize(1);
        assertThat(result.get(0).name()).isEqualTo("A");
        assertThat(result.get(0).books()).hasSize(3);
        assertThat(result.get(0).books()).extracting("name").containsExactlyInAnyOrder("책1", "책2", "책3");
        assertThat(result.get(0).books()).extracting("isReturn").containsExactlyInAnyOrder(false, false, true);
    }
}

File: src/test/kotlin/com/library/service/user/UserServiceTest.kt
package com.library.service.user

import com.library.domain.user.User
import com.library.domain.user.UserRepository
import com.library.dto.user.UserCreateRequest
import com.library.dto.user.UserUpdateRequest
import org.assertj.core.api.AssertionsForInterfaceTypes.assertThat
import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest

@SpringBootTest
class UserServiceTest @Autowired constructor(
    private val userService: UserService,
    private val userRepository: UserRepository,
) {
    @AfterEach
    fun clear() {
        userRepository.deleteAll()
    }

    @Test
    @DisplayName("유저 저장 테스트")
    fun saveUserTest() {
        // given
        val request = UserCreateRequest("유저1", null)

        // when
        userService.saveUser(request)

        // then
        val users = userRepository.findAll()
        assertThat(users).hasSize(1)
    }

    @Test
    @DisplayName("유저 리스트 조회 테스트")
    fun getUsersTest() {
        // given
        userRepository.saveAll(listOf(
            User.fixture("유저1", 20),
            User.fixture("유저2", null),
        ))

        // when
        val users = userService.getUsers()

        // then
        assertThat(users).hasSize(2)
        assertThat(users).extracting("name").containsExactlyInAnyOrder("유저1", "유저2")
        assertThat(users).extracting("age").containsExactlyInAnyOrder(20, null)
    }
    
    @Test
    @DisplayName("유저 이름 업데이트 테스트")
    fun updateUserNameTest() {
        // given
        val savedUser = userRepository.save(User.fixture("유저1"))

        // when
        userService.updateUserName(UserUpdateRequest(savedUser.id!!, "유저2"))
        
        // then
        val user = userRepository.findAll()[0]
        assertThat(user.name).isEqualTo("유저2")
    }

    @Test
    @DisplayName("유저 삭제 테스트")
    fun deleteUserTest() {
        // given
        userRepository.save(User.fixture("유저1"))

        // when
        userService.deleteUser("유저1")

        // then
        assertThat(userRepository.findAll()).isEmpty()
    }
}